### DISCUSSION TITLE
Is it possible to run periodic tasks?

### DISCUSSION THREAD

**Discussion by tksuns12** [serverpod/serverpod#345] [Q&A] [answered]
I'm trying to detect the expired data from the database and clean them once a day.

I can see that I can call the future calls after the server starts but without calling any endpoint, how should I execute some periodic task on the server?

**Comment by II11II** [comment]
Cron jobs?

**Reply by tksuns12** [reply]
Yes exactly. To put it simply, I want to know how to call `FutureCall`s repeatedly every certain time as soon as the server is on without triggering any endpoint.

**Comment by vlidholt** [comment, marked as answer]
You can create another future call at the end of your future call. If that makes sense? 

That being said, adding recurring calls would be a nice future feature.

**Reply by tksuns12** [reply]
Yeah that makes sense but I'm still curious of how to call that initial call for FutureCall without triggering any endpoint. This cron job should start itself after the server starts. Can I just unawait the server and execute some FutureCalls after?

**Reply by vlidholt** [reply]
Just place it after `await pod.start();` in your `server.dart` file. The `start` method doesn't block indefinitely.

E.g.

```dart
await pod.start();

await pod.futureCallWithDelay(callName, object, delay);
```

**Reply by tksuns12** [reply]
I misunderstood it totally. Thank you very much!

**Reply by gazialankus** [reply]
@vlidholt does this not start one future call with every restart/spawn of the server? My understanding is that our server can restart anytime or we can have multiple servers running. 

Is there a way to ensure a single future call is happening even if we have restarts or multiple servers running? 

I think at the moment the user needs to do
```
FutureCallEntry.db.findFirstRow(
  session,
  where: (t) => t.identifier.equals(identifier),
)
```
to check whether it's already scheduled. I think it would be nice to add a isFutureCallScheduled() function in FutureCallManager.

**Reply by gazialankus** [reply]
Also in main() we don't have a session so we have to create our own session. Gets quite hairy to ensure a single periodic FutureCall if I'm not mistaken.

**Reply by SandPod** [reply]
`isFutureCallScheduled ` is a good suggestion @gazialankus  - Would you mind filing an issue for that?

**Reply by gazialankus** [reply]
Ok I just did, thank you @SandPod https://github.com/serverpod/serverpod/issues/3928

As I was typing it I realized transactions around it would be nice as well 😅
